cpp/tests/collect_list/collect_list_test.cpp

```diff
@@ -0,0 +1,594 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_utilities.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/cudf_gtest.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <cudf/aggregation.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/rolling.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <src/rolling/rolling_detail.hpp>
+
+#include <thrust/iterator/constant_iterator.h>
+
+#include <algorithm>
+#include <vector>
+
+struct CollectListTest : public cudf::test::BaseFixture {
+};
+
+template <typename T>
+struct TypedCollectListTest : public CollectListTest {
+};
+
+using TypesForTest = cudf::test::
+  Concat<cudf::test::IntegralTypes, cudf::test::FloatingPointTypes, cudf::test::DurationTypes>;
+
+TYPED_TEST_CASE(TypedCollectListTest, TypesForTest);
+
+TYPED_TEST(TypedCollectListTest, BasicRollingWindow)
+{
+  using namespace cudf;
+  using namespace cudf::test;
+
+  using T = TypeParam;
+
+  auto const input_column = fixed_width_column_wrapper<T, int32_t>{10, 11, 12, 13, 14};
```

Comment:
I don't see any test case where `input_column` has nulls and resulting list will have null value.

Reply:
Good eye! There are a couple of combinations for which I'll need to add further tests.

In this case, I'll be adding tests for when the collected column has nulls, and for when the collected column is a nested column whose child has nulls. :/

I'll update here when I've added those.

Reply:
How to handle nulls ? Spark will skip the nulls per the code here https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala#L55

Reply:
> Spark will skip the nulls per the code

Hmm. Based on my tests, I thought Spark honours the null inputs. I must have been mistaken. Let me see if there is an easy way to support this.

Reply:
There is currently no good way to make the filtering behaviour configurable, at least with the `cudf::rolling_window()` APIs. @jrhemstad's advice on #7106 should allow it more cleanly.

Reply:
On examining at the behaviour of `COLLECT` in group-by-aggregation, it appears that null inputs are honoured in the output. This is congruent with the rest of CUDF, but at odds with the behaviour in Spark.

I can try adding a `null_policy` argument to `make_collect_aggregation()` while retaining the current behaviour as default.

There should be a way to post-process the child column to discard nulls. The null-mask calculation will remain unchanged (because it does not depend on the child column).

I am working on this right now. Will update here as soon as I have it running.

Reply:
I have spent the better part of today searching for a way to compute the gather map and offsets in one pass, taking into account both `min_periods` and null input columns. It does not look straightforward.
I think our only recourse is post processing, to filter out null entries from the gather map, and recomputing the offsets.

I should have an update shortly.

Reply:
Ok, I think supporting `null_policy::EXCLUDE` will need to be punted to post-release. I will handle this in a follow-on.

I will raise a CUDF issue for this and link it here.

@rgsl888prabhu, does that sound acceptable?

Reply:
(For the record, I am close to getting the list offset recomputation right. But it would be best to tackle this separately, with more breathing room.)